### PR TITLE
Implement inverse for rectangular polyconic projection

### PR DESCRIPTION
**Pull request by jbuonagurio**
As discussed with @kbevers on the mailing list and issue #420. This is a first attempt at the spherical inverse for the rectangular "war office" polyconic projection. It is heavily based on [s_inverse in PJ_poly.c](src/PJ_poly.c#L95-L118).

This is my first PR to proj.4 - any advice on test and documentation requirements is appreciated!

### CODE REVIEW

**Comment by kbevers** [comment]
Thanks John! It looks good at a glance, but I'll have to give it a closer look later.

In terms of tests, I would like you to add test data in `pj_rpoly_test()`. If you need an example for that have a look at the test in `PJ_aea.c`(or almost any other `PJ_*.c`-file). If I remember correctly, there is some test data in Yang, Snyder and Tobler. I suggest you use that to test you implementation.

It would be great to also update the documentation. Have a look at the Mercator docs for a template. I'm happy as long as you describe the maths behind the projection.

File: src/PJ_rpoly.c [hunk 1]
```diff
@@ -34,6 +38,35 @@ static XY s_forward (LP lp, PJ *P) {           /* Spheroidal, forward */
 }
 
 
+static LP s_inverse (XY xy, PJ *P) {           /* Spheroidal, inverse */
+    LP lp = {0.0,0.0};
+    struct pj_opaque *Q = P->opaque;
```
Comment (kbevers) [inline src/PJ_rpoly.c:43]: You don't seem to use Q anywhere in this function, so no need to assign it

**Review by kbevers — CHANGES_REQUESTED** [review]
I believe the code is functionally sane. The code in s_inverse is quite difficult to read, so I would like to see a cleaned up version of it, now that we're touching it anyway.

Test data for s_inverse is needed. 

Extra points for adding documentation of the rectangular polyconic but not strictly necessary for this PR to be merged.

File: src/PJ_rpoly.c (hunk 1 above)
Comment (jbuonagurio) [reply src/PJ_rpoly.c:43]: Will remove, I was using it in a previous iteration to access `Q->mode`.

File: src/PJ_rpoly.c [hunk 2]
```diff
@@ -34,6 +38,35 @@ static XY s_forward (LP lp, PJ *P) {           /* Spheroidal, forward */
...
+    double B, dphi, tp;
+    int i;
+
+    if (fabs(xy.y = P->phi0 + xy.y) <= TOL) {
```
[the remark this answers is not in the document]
Comment (jbuonagurio) [reply src/PJ_rpoly.c:47]: That was going to be my next comment. I basically tried to match the style of PJ_poly.c (american polyconic) for the sake of consistency, so you can see exactly how the two implementations are different, but I agree this is old-fashioned C and difficult to read. If we were compiling this with gcc -Wall it certainly wouldn't fly.

Other than the assignments in the if statements, I have several other issues with the code:
- `Q->fxa` and `Q->fxb` set in `PJ *PROJECTION()` and used in the forward calculations. It's confusing and makes it hard to see how the equations translate to code.
- I don't like the do while loop for the Newton-Raphson iteration.
- Should I replace the `I_ERROR` macro with `pj_ctx_set_errno`?

So if I make changes here for code readability, I feel like I should update pj_poly.c as well for consistency. Or do you prefer minimal changes to existing code? Question of philosophy...

File: src/PJ_rpoly.c [hunk 3]
```diff
@@ -63,7 +96,8 @@ PJ *PROJECTION(rpoly) {
     }
     P->es = 0.;
     P->fwd = s_forward;
-
+    P->inv = s_inverse;
+    
     return P;
 }
```
[the remark this answers is not in the document]
Comment (jbuonagurio) [reply src/PJ_rpoly.c:103]: Sure, I'll use their data for the self-tests. I'm already using it in ad-hoc `cs2cs` scripts for testing.

**Comment by jbuonagurio** [comment]
Kristian, thank you for the code review. I've added comments to the above. I should have some time this weekend to finalize the PR.

File: src/PJ_rpoly.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (kbevers) [reply src/PJ_rpoly.c:47]: > Q->fxa and Q->fxb set in PJ *PROJECTION() and used in the forward calculations. It's confusing and makes it hard to see how the equations translate to code.

I agree, it is very difficult to understand how the code does what the equations says!

> I don't like the do while loop for the Newton-Raphson iteration.

Me too.

> Should I replace the I_ERROR macro with pj_ctx_set_errno?

Don't bother. I think it is inconsistently used in the code and should probably be fixed at some across the whole code base in a dedicated effort. I know @busstoptaktik has been talking about getting rid of the ERROR macros.

> So if I make changes here for code readability, I feel like I should update pj_poly.c as well for consistency. Or do you prefer minimal changes to existing code? Question of philosophy...

That's always a tough one. As long as the changes are backed up by tests I think it adds value to the code. Go ahead :)

**Comment by jbuonagurio** [comment]
I have a bit more work to do on this. In addition to the readability changes discussed above, I need to implement the forward and inverse on the ellipsoid to match Yang's test data - he is using the Krassovsky (1940) ellipsoid (+ellps=krass) which seems to be popular in China.

I've done the math on paper (this is again very similar to the American polyconic), just need to finish translating to C. Probably just a few more days.

**Comment by kbevers** [comment]
Thanks for the update. Very cool to get the projection on the ellipsoid as well!

**Comment by kbevers** [comment]
@jbuonagurio How are things going with this PR?

**Comment by busstoptaktik** [comment]
@jbuonagurio and @kbevers - this PR has been quiet for (almost) a year. Should it be closed, or can it be merged as-is leaving out the ellipsoidal version?

**Comment by kbevers** [comment]
Unless @jbuonagurio intends to finish this I think it should be closed. It is mostly finished if I remember correctly, but there is no test for the new inverse implementation. I don't have (easy) access to the source material where test data can be found, otherwise I'd finish it myself.

**Comment by jbuonagurio** [comment]
@busstoptaktik and @kbevers - I would like to finish this, especially as you were so kind to provide a code review. Apologies for the lack of activity on this PR - I got tied up in 'company work'. I can also scan/digitize the relevant test data from Yang et al. Give me about a week to free up some time.

**Comment by kbevers** [comment]
> I would like to finish this

@jbuonagurio that sounds great. Things have changed a bit since your last commit, so parts of the code will probably have to be reworked. Let med know if you need guidance on that.

**Comment by busstoptaktik** [comment]
I think we should close this for now - @jbuonagurio you can reopen once you find time to finish this almost-done work, but it should probably not sit in the open queue indicating that it is in active development.

**Pull request closed without merging**